react/features/vad-reporter/VADReportingService.js

```diff
@@ -0,0 +1,247 @@
+                    logger.error(`Create TrackVADEmitter promise failed with ${outcome.reason}`);
+
+                    rejectedEmitterPromiseArray.push(outcome);
+                }
+            });
+
+            // Check if there were any rejected promises and clear the already created ones list.
+            if (rejectedEmitterPromiseArray.length > 0) {
+
+                logger.error('Cleaning up remaining VADDeviceContext, due to create fail!');
+
+                vadContextArray.forEach(context => {
+                    context.vadEmitter.destroy();
+                });
+
+                // Reject create promise if one emitter failed to instantiate, we might one just ignore it,
+                // leaving it like this for now
+                throw new Error('Create VADReportingService failed due to TrackVADEmitter creation issues!');
+            }
+
+            vadReportingService._setVADContextArray(vadContextArray);
+            vadReportingService._startPublish();
+
+            return vadReportingService;
+        });
+    }
+
+    /**
+     * Destroy TrackVADEmitters and clear the context map.
+     *
+     * @returns {void}
+     */
+    _clearContextMap() {
+        this._contextMap.forEach(vadContext => vadContext.vadEmitter.destroy());
+        this._contextMap.clear();
+    }
+
+    /**
+     * Set the watched device contexts.
+     *
+     * @param {Array<VADDeviceContext>} vadContextArray - List of mics.
+     * @returns {void}
+     */
+    _setVADContextArray(vadContextArray: Array<VADDeviceContext>): void {
+        vadContextArray.forEach((vadContext: Object) => {
+            this._contextMap.set(vadContext.deviceInfo.deviceId, vadContext);
+        });
+    }
+
+    /**
+     * Start the setInterval reporting process.
+     *
+     * @returns {void}.
+     */
+    _startPublish() {
+        logger.log('VADReportingService started publishing.');
+        this._intervalId = setInterval(() => {
+            this._reportVadScore();
+        }, this._intervalDelay);
+    }
+
+    /**
+     * Function called at set interval with selected compute. The result will be published on the set callback.
+     *
+     * @returns {void}
+     */
+    _reportVadScore() {
+        const vadComputeScoreArray = [];
+        const computeTimestamp = Date.now();
+
+        // Go through each device and compute cumulated VAD score.
+        this._contextMap.forEach((vadContext, deviceId) => {
+            const nrOfVADScores = vadContext.scoreArray.length;
+            let vadSum = 0;
+
+            vadContext.scoreArray.forEach(vadScore => {
+                vadSum += vadScore.score;
+            });
+
+            // TODO For now we just calculate the average score for each device, more compute algorithms will be added.
+            const avgVAD = vadSum / nrOfVADScores;
+
+            vadContext.scoreArray = [];
+
+            vadComputeScoreArray.push({
+                timestamp: computeTimestamp,
+                score: avgVAD,
```

Comment:
maybe let's call it `avgVAD` instead of `score` to be more explicit?

Reply:
I was planning on adding different methods for calculating the score along side AVG, that's why the generic name.

Reply:
Ok